orchestrator/utils/redis.py

```diff
@@ -152,12 +161,24 @@ async def subscriber(self, *channels: str) -> AsyncGenerator[PubSub, None]:
         Automatically unsubscribes and releases the connection afterwards.
         """
         pubsub = self.client.pubsub(ignore_subscribe_messages=True)
+
+        async def do_teardown() -> None:
+            if not pubsub.subscribed:
+                return
+
+            await pubsub.unsubscribe(*channels)
+            await pubsub.aclose()  # type: ignore[attr-defined]
+
         try:
             await pubsub.subscribe(*channels)
             yield pubsub
+        except get_cancelled_exc_class():
+            # https://anyio.readthedocs.io/en/3.x/cancellation.html#finalization
+            with CancelScope(shield=True):
+                await do_teardown()
+            raise
         finally:
-            await pubsub.unsubscribe(*channels)
-            await pubsub.aclose()  # type: ignore[attr-defined]
+            await do_teardown()
```

Comment:
good to know of the finalization cancel scope, is this finally also nessecary?

Reply:
Yes. It's recommended practice to use the`try/finally` construct in contextmanagers, in case we leave the context for any other reason. The cancel scope is just an addition to prevent anyio from throwing the async task into the void :)